enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,568 @@
+
+As with the installer, the vSphere [machine reconciler](https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/vsphere/reconciler.go#L745-L755) 
+will pass the static IP configuration via the `guestinfo.afterburn.initrd.network-kargs` extraconfig parameter.  
+
+
+### Day 2 Static IP Configuration
+
+Nodes being added to a cluster may be configured via an `nmstate` IP configuration or default to DHCP.  The networking configuration of a node/machine is immutable after creation. The vSphere machine API machine controller will apply the nmstate configuration when the associated VM is cloned.
+
+`machinesets` will be supported through the creation of a user-created custom controller.  This custom controller will leverage machine lifecycle hooks to
+provide IP configuration to machines descending from `machinesets` with `machine` annotated with the appropriate lifecycle hook.
+
+#### Changes Required
+
+##### Installer
+
+1. Modify the `install-config.yaml` vSphere platform specification to support nmstate configuration. The nmstate configuration is intended to be structurally compatible with the [nmstate API](https://nmstate.io/) but will omit fields that do not map logically to installation of a cluster.
+
+~~~go
+// Hosts defines `Host` configurations to be applied to nodes deployed by the installer
+type Hosts struct {
+  // Bootstrap IP configuration for the bootstrap node. If omitted, node will default to DHCP.
+  // +optional
+  Bootstrap *Host `json: "bootstrap,omitempty"`
+
+  // ControlPlane slice of IP configurations for the control plane nodes. If empty, nodes will default to DHCP.
+  // +optional
+  ControlPlane []Host `json: "controlPlane,omitempty"`
+
+  // Compute slice of IP configurations for the compute nodes. If empty, nodes will default to DHCP.
+  // +optional
+  Compute []Host `json: "compute,omitempty"`
+}
+
+// Host defines the IP configuration to be applied for a node deployed by the installer
+type Host struct {  
+  // FailureDomain refers to the name of a FailureDomain as described in https://github.com/openshift/enhancements/blob/master/enhancements/installer/vsphere-ipi-zonal.md
+  // +optional
+  FailureDomain string `json: "failureDomain"`
+
+  // NetworkConfig IP configuration to be applied
+  // +kubebuilder:validation:Required
+  NetworkConfig machineapiv1beta1.NetworkConfig string `json: "networkConfig"` 
+}
+
+// NetworkConfig Config specifies an IP configuration to be applied upon cloning.
+type NetworkConfig struct {
+  // Interfaces slice of interfaces to be configured.	
+  // +kubebuilder:validation:MaxItems=1
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:Required
+  Interfaces []Interface `json:"interfaces"`
+
+  // DnsResolver defines nameservers to be applied to a network interface
+  // interfaces.
+  // +kubebuilder:validation:Required
+  DnsResolver DnsResolver `json:"dns-resolver"`
```

Comment:
the casing convention here is that used by nmstate https://nmstate.io/examples.html#dns .

Reply:
I'm certainly open to discussion on this if you think we should deviate from the nmstate API

Reply:
I think the idea of using the nmstate names for the JSON representation (without actually accepting all of nmstate or relying on nmstatectl for anything) was a good one.
The golang field names can be changed to fit local coding conventions.

Reply:
In general our API conventions are to be consistent within our own APIs rather than consistent with third party APIs such as cloud provider APIs. This allows us to create a familiar feel across the entire product.
Patricks suggestion is correct

Reply:
Thanks all for the feedback on this.  I'll make those changes.

Reply:
Changing the JSON names breaks compatibility with all of the existing uses of NMState in OpenShift. This is going to be very confusing to users and if it is a hard requirement then I lean more heavily toward not representing this in the API at all and just treating the config as a JSON blob to be processed by an NMState-compatible translation layer, similar to how it works in baremetal and assisted-installer.

Reply:
NMState is a first-party API. It didn't come from a cloud provider, Red Hat invented it and it's a standard way that administrators of RHEL in general configure their networks, so using it creates a familiar feel across the entire portfolio.
Using a subset of NMState would also be consistent with our own APIs where we've already embedded it in other places in the install-config (as well as in CRDs like NodeNetworkConfigurationPolicy).
Making up a new thing that is not nmstate-like would be far more inconsistent IMHO.

Reply:
I think I'll raise this as a topic for discussion at the next cluster lifecycle arch call.   @cybertron @zaneb that makes a lot of sense.  My original intent here was just to include a subset of the nmstate API.  But given that it clashes with OpenShift API conventions, I'm not quite sure how to disposition this other than doing what baremetal/AI did and just passing the nmstate along as a json blob.  Thanks for all the feedback on this.

Reply:
We have many examples where we have deviated in the past, and as API approvers, do not accept "we did it already so it's ok now". We have API review for a specific reason, in the past, this hasn't always happened or different reviewers have had different bars. The bar is now set and this must match OpenShift API conventions if you want this to merge.

The API surface of OpenShift is much bigger than just the NMState implementations in other areas of the API, for this to be accepted it should match the majority of the OpenShift API surface, and not necessarily the NMState API in another area of the product which I'm guessing hasn't gone through API review

Our API [FAQs](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#faqs) are relevant here

Reply:
Thanks @JoelSpeed .  I'm certainly not trying to work around the review process here.  I'm just trying to take in to account the feedback and precedent we have followed on a similar enhancement for nmstate.  The currently committed changes should represent the changes necessary to adhere to the OpenShift API guidelines.